// @azure/arm-containerregistry/dist-esm/src/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link ImportMode} that the service accepts. */
export var KnownImportMode;
(function (KnownImportMode) {
    KnownImportMode["NoForce"] = "NoForce";
    KnownImportMode["Force"] = "Force";
})(KnownImportMode || (KnownImportMode = {}));
/** Known values of {@link ConnectionStatus} that the service accepts. */
export var KnownConnectionStatus;
(function (KnownConnectionStatus) {
    KnownConnectionStatus["Approved"] = "Approved";
    KnownConnectionStatus["Pending"] = "Pending";
    KnownConnectionStatus["Rejected"] = "Rejected";
    KnownConnectionStatus["Disconnected"] = "Disconnected";
})(KnownConnectionStatus || (KnownConnectionStatus = {}));
/** Known values of {@link ActionsRequired} that the service accepts. */
export var KnownActionsRequired;
(function (KnownActionsRequired) {
    KnownActionsRequired["None"] = "None";
    KnownActionsRequired["Recreate"] = "Recreate";
})(KnownActionsRequired || (KnownActionsRequired = {}));
/** Known values of {@link ProvisioningState} that the service accepts. */
export var KnownProvisioningState;
(function (KnownProvisioningState) {
    KnownProvisioningState["Creating"] = "Creating";
    KnownProvisioningState["Updating"] = "Updating";
    KnownProvisioningState["Deleting"] = "Deleting";
    KnownProvisioningState["Succeeded"] = "Succeeded";
    KnownProvisioningState["Failed"] = "Failed";
    KnownProvisioningState["Canceled"] = "Canceled";
})(KnownProvisioningState || (KnownProvisioningState = {}));
/** Known values of {@link CreatedByType} that the service accepts. */
export var KnownCreatedByType;
(function (KnownCreatedByType) {
    KnownCreatedByType["User"] = "User";
    KnownCreatedByType["Application"] = "Application";
    KnownCreatedByType["ManagedIdentity"] = "ManagedIdentity";
    KnownCreatedByType["Key"] = "Key";
})(KnownCreatedByType || (KnownCreatedByType = {}));
/** Known values of {@link LastModifiedByType} that the service accepts. */
export var KnownLastModifiedByType;
(function (KnownLastModifiedByType) {
    KnownLastModifiedByType["User"] = "User";
    KnownLastModifiedByType["Application"] = "Application";
    KnownLastModifiedByType["ManagedIdentity"] = "ManagedIdentity";
    KnownLastModifiedByType["Key"] = "Key";
})(KnownLastModifiedByType || (KnownLastModifiedByType = {}));
/** Known values of {@link SkuName} that the service accepts. */
export var KnownSkuName;
(function (KnownSkuName) {
    KnownSkuName["Classic"] = "Classic";
    KnownSkuName["Basic"] = "Basic";
    KnownSkuName["Standard"] = "Standard";
    KnownSkuName["Premium"] = "Premium";
})(KnownSkuName || (KnownSkuName = {}));
/** Known values of {@link SkuTier} that the service accepts. */
export var KnownSkuTier;
(function (KnownSkuTier) {
    KnownSkuTier["Classic"] = "Classic";
    KnownSkuTier["Basic"] = "Basic";
    KnownSkuTier["Standard"] = "Standard";
    KnownSkuTier["Premium"] = "Premium";
})(KnownSkuTier || (KnownSkuTier = {}));
/** Known values of {@link DefaultAction} that the service accepts. */
export var KnownDefaultAction;
(function (KnownDefaultAction) {
    KnownDefaultAction["Allow"] = "Allow";
    KnownDefaultAction["Deny"] = "Deny";
})(KnownDefaultAction || (KnownDefaultAction = {}));
/** Known values of {@link Action} that the service accepts. */
export var KnownAction;
(function (KnownAction) {
    KnownAction["Allow"] = "Allow";
})(KnownAction || (KnownAction = {}));
/** Known values of {@link PolicyStatus} that the service accepts. */
export var KnownPolicyStatus;
(function (KnownPolicyStatus) {
    KnownPolicyStatus["Enabled"] = "enabled";
    KnownPolicyStatus["Disabled"] = "disabled";
})(KnownPolicyStatus || (KnownPolicyStatus = {}));
/** Known values of {@link TrustPolicyType} that the service accepts. */
export var KnownTrustPolicyType;
(function (KnownTrustPolicyType) {
    KnownTrustPolicyType["Notary"] = "Notary";
})(KnownTrustPolicyType || (KnownTrustPolicyType = {}));
/** Known values of {@link ExportPolicyStatus} that the service accepts. */
export var KnownExportPolicyStatus;
(function (KnownExportPolicyStatus) {
    KnownExportPolicyStatus["Enabled"] = "enabled";
    KnownExportPolicyStatus["Disabled"] = "disabled";
})(KnownExportPolicyStatus || (KnownExportPolicyStatus = {}));
/** Known values of {@link EncryptionStatus} that the service accepts. */
export var KnownEncryptionStatus;
(function (KnownEncryptionStatus) {
    KnownEncryptionStatus["Enabled"] = "enabled";
    KnownEncryptionStatus["Disabled"] = "disabled";
})(KnownEncryptionStatus || (KnownEncryptionStatus = {}));
/** Known values of {@link PublicNetworkAccess} that the service accepts. */
export var KnownPublicNetworkAccess;
(function (KnownPublicNetworkAccess) {
    KnownPublicNetworkAccess["Enabled"] = "Enabled";
    KnownPublicNetworkAccess["Disabled"] = "Disabled";
})(KnownPublicNetworkAccess || (KnownPublicNetworkAccess = {}));
/** Known values of {@link NetworkRuleBypassOptions} that the service accepts. */
export var KnownNetworkRuleBypassOptions;
(function (KnownNetworkRuleBypassOptions) {
    KnownNetworkRuleBypassOptions["AzureServices"] = "AzureServices";
    KnownNetworkRuleBypassOptions["None"] = "None";
})(KnownNetworkRuleBypassOptions || (KnownNetworkRuleBypassOptions = {}));
/** Known values of {@link ZoneRedundancy} that the service accepts. */
export var KnownZoneRedundancy;
(function (KnownZoneRedundancy) {
    KnownZoneRedundancy["Enabled"] = "Enabled";
    KnownZoneRedundancy["Disabled"] = "Disabled";
})(KnownZoneRedundancy || (KnownZoneRedundancy = {}));
/** Known values of {@link RegistryUsageUnit} that the service accepts. */
export var KnownRegistryUsageUnit;
(function (KnownRegistryUsageUnit) {
    KnownRegistryUsageUnit["Count"] = "Count";
    KnownRegistryUsageUnit["Bytes"] = "Bytes";
})(KnownRegistryUsageUnit || (KnownRegistryUsageUnit = {}));
/** Known values of {@link WebhookStatus} that the service accepts. */
export var KnownWebhookStatus;
(function (KnownWebhookStatus) {
    KnownWebhookStatus["Enabled"] = "enabled";
    KnownWebhookStatus["Disabled"] = "disabled";
})(KnownWebhookStatus || (KnownWebhookStatus = {}));
/** Known values of {@link WebhookAction} that the service accepts. */
export var KnownWebhookAction;
(function (KnownWebhookAction) {
    KnownWebhookAction["Push"] = "push";
    KnownWebhookAction["Delete"] = "delete";
    KnownWebhookAction["Quarantine"] = "quarantine";
    KnownWebhookAction["ChartPush"] = "chart_push";
    KnownWebhookAction["ChartDelete"] = "chart_delete";
})(KnownWebhookAction || (KnownWebhookAction = {}));
/** Known values of {@link OS} that the service accepts. */
export var KnownOS;
(function (KnownOS) {
    KnownOS["Windows"] = "Windows";
    KnownOS["Linux"] = "Linux";
})(KnownOS || (KnownOS = {}));
/** Known values of {@link RunStatus} that the service accepts. */
export var KnownRunStatus;
(function (KnownRunStatus) {
    KnownRunStatus["Queued"] = "Queued";
    KnownRunStatus["Started"] = "Started";
    KnownRunStatus["Running"] = "Running";
    KnownRunStatus["Succeeded"] = "Succeeded";
    KnownRunStatus["Failed"] = "Failed";
    KnownRunStatus["Canceled"] = "Canceled";
    KnownRunStatus["Error"] = "Error";
    KnownRunStatus["Timeout"] = "Timeout";
})(KnownRunStatus || (KnownRunStatus = {}));
/** Known values of {@link RunType} that the service accepts. */
export var KnownRunType;
(function (KnownRunType) {
    KnownRunType["QuickBuild"] = "QuickBuild";
    KnownRunType["QuickRun"] = "QuickRun";
    KnownRunType["AutoBuild"] = "AutoBuild";
    KnownRunType["AutoRun"] = "AutoRun";
})(KnownRunType || (KnownRunType = {}));
/** Known values of {@link Architecture} that the service accepts. */
export var KnownArchitecture;
(function (KnownArchitecture) {
    KnownArchitecture["Amd64"] = "amd64";
    KnownArchitecture["X86"] = "x86";
    KnownArchitecture["ThreeHundredEightySix"] = "386";
    KnownArchitecture["Arm"] = "arm";
    KnownArchitecture["Arm64"] = "arm64";
})(KnownArchitecture || (KnownArchitecture = {}));
/** Known values of {@link Variant} that the service accepts. */
export var KnownVariant;
(function (KnownVariant) {
    KnownVariant["V6"] = "v6";
    KnownVariant["V7"] = "v7";
    KnownVariant["V8"] = "v8";
})(KnownVariant || (KnownVariant = {}));
/** Known values of {@link TaskStatus} that the service accepts. */
export var KnownTaskStatus;
(function (KnownTaskStatus) {
    KnownTaskStatus["Disabled"] = "Disabled";
    KnownTaskStatus["Enabled"] = "Enabled";
})(KnownTaskStatus || (KnownTaskStatus = {}));
/** Known values of {@link StepType} that the service accepts. */
export var KnownStepType;
(function (KnownStepType) {
    KnownStepType["Docker"] = "Docker";
    KnownStepType["FileTask"] = "FileTask";
    KnownStepType["EncodedTask"] = "EncodedTask";
})(KnownStepType || (KnownStepType = {}));
/** Known values of {@link BaseImageDependencyType} that the service accepts. */
export var KnownBaseImageDependencyType;
(function (KnownBaseImageDependencyType) {
    KnownBaseImageDependencyType["BuildTime"] = "BuildTime";
    KnownBaseImageDependencyType["RunTime"] = "RunTime";
})(KnownBaseImageDependencyType || (KnownBaseImageDependencyType = {}));
/** Known values of {@link TriggerStatus} that the service accepts. */
export var KnownTriggerStatus;
(function (KnownTriggerStatus) {
    KnownTriggerStatus["Disabled"] = "Disabled";
    KnownTriggerStatus["Enabled"] = "Enabled";
})(KnownTriggerStatus || (KnownTriggerStatus = {}));
/** Known values of {@link SourceControlType} that the service accepts. */
export var KnownSourceControlType;
(function (KnownSourceControlType) {
    KnownSourceControlType["Github"] = "Github";
    KnownSourceControlType["VisualStudioTeamService"] = "VisualStudioTeamService";
})(KnownSourceControlType || (KnownSourceControlType = {}));
/** Known values of {@link TokenType} that the service accepts. */
export var KnownTokenType;
(function (KnownTokenType) {
    KnownTokenType["PAT"] = "PAT";
    KnownTokenType["OAuth"] = "OAuth";
})(KnownTokenType || (KnownTokenType = {}));
/** Known values of {@link SourceTriggerEvent} that the service accepts. */
export var KnownSourceTriggerEvent;
(function (KnownSourceTriggerEvent) {
    KnownSourceTriggerEvent["Commit"] = "commit";
    KnownSourceTriggerEvent["Pullrequest"] = "pullrequest";
})(KnownSourceTriggerEvent || (KnownSourceTriggerEvent = {}));
/** Known values of {@link BaseImageTriggerType} that the service accepts. */
export var KnownBaseImageTriggerType;
(function (KnownBaseImageTriggerType) {
    KnownBaseImageTriggerType["All"] = "All";
    KnownBaseImageTriggerType["Runtime"] = "Runtime";
})(KnownBaseImageTriggerType || (KnownBaseImageTriggerType = {}));
/** Known values of {@link UpdateTriggerPayloadType} that the service accepts. */
export var KnownUpdateTriggerPayloadType;
(function (KnownUpdateTriggerPayloadType) {
    KnownUpdateTriggerPayloadType["Default"] = "Default";
    KnownUpdateTriggerPayloadType["Token"] = "Token";
})(KnownUpdateTriggerPayloadType || (KnownUpdateTriggerPayloadType = {}));
/** Known values of {@link SourceRegistryLoginMode} that the service accepts. */
export var KnownSourceRegistryLoginMode;
(function (KnownSourceRegistryLoginMode) {
    KnownSourceRegistryLoginMode["None"] = "None";
    KnownSourceRegistryLoginMode["Default"] = "Default";
})(KnownSourceRegistryLoginMode || (KnownSourceRegistryLoginMode = {}));
/** Known values of {@link SecretObjectType} that the service accepts. */
export var KnownSecretObjectType;
(function (KnownSecretObjectType) {
    KnownSecretObjectType["Opaque"] = "Opaque";
    KnownSecretObjectType["Vaultsecret"] = "Vaultsecret";
})(KnownSecretObjectType || (KnownSecretObjectType = {}));
