// @azure/container-registry/dist-esm/src/models.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
/**
 * Defines known cloud audiences for Azure Container Registry.
 */
export var KnownContainerRegistryAudience;
(function (KnownContainerRegistryAudience) {
    /** Azure China */
    KnownContainerRegistryAudience["AzureResourceManagerChina"] = "https://management.chinacloudapi.cn";
    /** Azure Gemany */
    KnownContainerRegistryAudience["AzureResourceManagerGermany"] = "https://management.microsoftazure.de";
    /** Azure Government */
    KnownContainerRegistryAudience["AzureResourceManagerGovernment"] = "https://management.usgovcloudapi.net";
    /** Azure Public Cloud */
    KnownContainerRegistryAudience["AzureResourceManagerPublicCloud"] = "https://management.azure.com";
})(KnownContainerRegistryAudience || (KnownContainerRegistryAudience = {}));
/**
 * Defines known artifact architectures that the service supports.
 */
export var KnownArtifactArchitecture;
(function (KnownArtifactArchitecture) {
    /** i386 */
    KnownArtifactArchitecture["I386"] = "386";
    /** AMD64 */
    KnownArtifactArchitecture["Amd64"] = "amd64";
    /** ARM */
    KnownArtifactArchitecture["Arm"] = "arm";
    /** ARM64 */
    KnownArtifactArchitecture["Arm64"] = "arm64";
    /** MIPS */
    KnownArtifactArchitecture["Mips"] = "mips";
    /** MIPSLE */
    KnownArtifactArchitecture["MipsLe"] = "mipsle";
    /** MIPS64 */
    KnownArtifactArchitecture["Mips64"] = "mips64";
    /** MIPS64LE */
    KnownArtifactArchitecture["Mips64Le"] = "mips64le";
    /** PPC64 */
    KnownArtifactArchitecture["Ppc64"] = "ppc64";
    /** PPC64LE */
    KnownArtifactArchitecture["Ppc64Le"] = "ppc64le";
    /** RISCv64 */
    KnownArtifactArchitecture["RiscV64"] = "riscv64";
    /** s390x */
    KnownArtifactArchitecture["S390X"] = "s390x";
    /** Wasm */
    KnownArtifactArchitecture["Wasm"] = "wasm";
})(KnownArtifactArchitecture || (KnownArtifactArchitecture = {}));
/**
 * Defines known artifact platform's operating systems that the service supports.
 */
export var KnownArtifactOperatingSystem;
(function (KnownArtifactOperatingSystem) {
    /** Aix */
    KnownArtifactOperatingSystem["Aix"] = "aix";
    /** Android */
    KnownArtifactOperatingSystem["Android"] = "android";
    /** Darwin */
    KnownArtifactOperatingSystem["Darwin"] = "darwin";
    /** Dragonfly */
    KnownArtifactOperatingSystem["Dragonfly"] = "dragonfly";
    /** FreeBSD */
    KnownArtifactOperatingSystem["FreeBsd"] = "freebsd";
    /** Illumos */
    KnownArtifactOperatingSystem["Illumos"] = "illumos";
    /** iOS */
    KnownArtifactOperatingSystem["iOS"] = "ios";
    /** JS */
    KnownArtifactOperatingSystem["JS"] = "js";
    /** Linux */
    KnownArtifactOperatingSystem["Linux"] = "linux";
    /** NetBSD */
    KnownArtifactOperatingSystem["NetBsd"] = "netbsd";
    /** OpenBSD */
    KnownArtifactOperatingSystem["OpenBsd"] = "openbsd";
    /** Plan9 */
    KnownArtifactOperatingSystem["Plan9"] = "plan9";
    /** Solaris */
    KnownArtifactOperatingSystem["Solaris"] = "solaris";
    /** Windows */
    KnownArtifactOperatingSystem["Windows"] = "windows";
})(KnownArtifactOperatingSystem || (KnownArtifactOperatingSystem = {}));
