// @azure/container-registry/dist-esm/src/transformations.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
export function toManifestWritableProperties(from) {
    // don't return unwanted properties, namely `quarantineState` and `quarantineDetails`
    return from
        ? {
            canDelete: from.canDelete,
            canList: from.canList,
            canRead: from.canRead,
            canWrite: from.canWrite,
        }
        : undefined;
}
export function toArtifactManifestProperties(from, repositoryName, registryLoginServer) {
    var _a, _b, _c, _d;
    return Object.assign({ registryLoginServer,
        repositoryName, digest: from.digest, sizeInBytes: from.size, createdOn: from.createdOn, lastUpdatedOn: from.lastUpdatedOn, architecture: (_a = from.architecture) !== null && _a !== void 0 ? _a : undefined, operatingSystem: (_b = from.operatingSystem) !== null && _b !== void 0 ? _b : undefined, relatedArtifacts: (_c = from.relatedArtifacts) !== null && _c !== void 0 ? _c : [], tags: (_d = from.tags) !== null && _d !== void 0 ? _d : [] }, toManifestWritableProperties(from));
}
export function toServiceTagOrderBy(orderBy) {
    return orderBy === "LastUpdatedOnAscending"
        ? "timeasc"
        : orderBy === "LastUpdatedOnDescending"
            ? "timedesc"
            : undefined;
}
export function toServiceManifestOrderBy(orderBy) {
    return orderBy === "LastUpdatedOnAscending"
        ? "timeasc"
        : orderBy === "LastUpdatedOnDescending"
            ? "timedesc"
            : undefined;
}
